Riiid 인터뷰 과제 PR

jinseoplee opened the pull request:
# Riiid 인터뷰 과제 PR


## 구동 방법

``` bash
# 프로젝트 클론 후
cd server && npm install

# 노드 서버 실행 후 http://localhost:3000/ 으로 접속 
# (quiz의 빌드파일을 server폴더로 떨어지게 만들어서 server만 실행하면 됩니다.)
npm start
```
## 사용 기술
server
* node, express

client(quiz)
* Vue
* SCSS

## 작업 내용

server 프로젝트와 quiz 프로젝트로 나눠서 작업 하였습니다.

### server

1. express를 이용하였으며 tesk_container.json을 읽어 '/quiz' 로 get 요청이 들어오면 데이터를 보내줍니다.

### client (quiz)

* vuex action에서 api 요청을 하고 utils/ 의 quiz-parser.js 로 데이터를 파싱하여 간결하게 만듭니다.(데이터를 제대로 사용한건지 아직도 잘 모르겠어요...)

* 파싱된 데이터로 UI를 렌더링 합니다. 데이터는 아래와 같이 사용되었습니다.

  - passageBox : 토익 문제의 article을 구성
  - passageBox > root : article 가장 상단 구문
  - passageBox > passage : article 본문
  - passageBox > passageTranslations : 문제 풀기 완료 후 리뷰에서 사용
  - passageBox > vocabularies : 문제 풀기 완료 후 리뷰에서 사용
  - questions : 문제 UI 생성에 사용

* 퀴즈 데이터와 퀴즈 진행 상황에 대한 데이터는 state에 담아 관리하였습니다.
* UI 플로우
  - 퀴즈는 Multi-step form으로 구성되어있습니다.
  - 보기중 하나라도 선택해야 정답을 확인 할 수 있습니다.
  - 정답 확인을 해야 다음 단계로 넘어갈 수 있습니다.
  - 최종 스텝까지 완료하면 "다시 풀기" 와 "리뷰" 버튼이 활성호 됩니다.
  - "다시 풀기"를 클릭하면 처음 스텝 부터 재 시작됩니다.
  - "리뷰" 를 클릭하면 본문의 해석본과 단어를 보여줍니다.

## 질문
일단 PR이 늦은점 정말 죄송합니다.
처음에는 별거 아닌거 같았는데 별거여서 조금 놀랐습니다. 데이터가 생각보다 복잡하더라구요..:) 그래도 재밌었습니다.
다름이 아니라 데이터중에 paragraph 아래 chunk가 항상 존재했는데 배열로 들어가 있더라구요. 저는 한 개의 paragraph당 한개의 chunk를 바라보는 1:1 관계로 더 많은 chunk가 필요하면 paragraph를 추가하는 식으로 봤었는데.. 혹시 한 개의 paragraph에서 여러 chunk를 바라보고 있는 사례가 있을까요??

zittto committed:
브런치 assignment 를 생성한다

server : express api 서버
quiz : vue.js 프로젝트

zittto committed:
quiz parser 추가

한 paragraph당 chunk가 한 개 이상 존재 할 수 있는 경우가 존재 하는가, 만약 존재한다면 paragraph을 추가하고 chunk와는 1:1 관계를 유지하는게 더 간결한 데이터가 되지 않을까 하는 생각.

quiz parser를 통해 데이터를 파싱하고 vuex state에 저장하는 로직도 추가

zittto committed:
passages와 quiz 렌더링

zittto committed:
UI 작업 완료

zittto committed:
font 적용 및 영어단어 정렬

shKwonRiiid commented:
@jinseoplee 
리뷰가 늦어져서 죄송합니다 🙇 
다음주 화요일까지 리뷰 남기겠습니다 🙏

vchoisk commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+  return head.concat(str).concat(tail);
+};
+
+const extract = {
```
함수별로 두지 않고 따로 extract object로 정의 하신 이유가 있나요?

jinseoplee replied:
퀴즈의 요소들을 추출하는 함수들을 묶어서 관리하고 싶어서 extract object 로 정의하였습니다

vchoisk commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+  question_area(qa) {
+    const paragraphArray = [];
+    if (qa.children) {
+      for (const child of qa.children) {
```
higher order function과 loop을 사용하는 기준이 있으신가요?

jinseoplee replied:
현재 데이터가 question_area > [ paragraph > [ chunk_ref ] ]와 같은 형식으로 이루어져 있는데, question_area 아래에 paragraph말고 다른 형태의 데이터가 올 수도 있지 않을까 하는 생각에 higher order function을 사용하였고, 데이터가 배열로 처리되어 있어서 더 많은 child가 추가 될 수 있을수도 있다고 생각하여 loop를 사용하였습니다.

vchoisk commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+  },
+};
+
+const QuizParser = (quiz) => {
```
함수명 정의시 대소문자 사용하는 기준이 무엇인가요?

jinseoplee replied:
아.. 함수명은 CamelCase로 통일하는 편인데 실수가 있었던것 같습니다.. 이 부분은 수정하도록 하겠습니다

jinseoplee replied:
해당 부분 수정하였습니다!
be27a04a0d264391dff908f6dfa246757f864c75

shKwonRiiid commented on quiz/src/components/Training.vue:
```diff
@@ -0,0 +1,255 @@
+      const result = this.getQuiz.passageBox.passages.filter((item) => {
+        return item.text && item.text.trim();
+      });
+      return result;
```
변수선언 하지않고

바로 return 해도 되지 않을까요?

jinseoplee replied:
음 넵 그렇게 하는게 더 간편할것 같네요. 처음에 lint에서 걸렸었던 것 같은데 제가 잘못 확인한것 같습니다.
저 방식으로 변경해보았습니다.
9863ae48030bae52d4c8daba9b8b0fde3e564746

vchoisk commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+};
+
+const QuizParser = (quiz) => {
+  extract.setChunkMap(quiz.chunk_map);
```
quiz를 destructuring 사용해서 
```
({name, type, preview, chunk_map}) => {
  ...
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

jinseoplee replied:
음 확실히 destructuring 을 사용해서 매개변수를 받는게 함수가 더 깔끔해지네요! 해당 방법으로 적용해보도록 하겠습니다!

jinseoplee replied:
destructuring을 적용해보니 더 깔끔한것 같습니다.
be27a04a0d264391dff908f6dfa246757f864c75

juyeong1260 commented on quiz/src/store/actions.js:
```diff
@@ -0,0 +1,27 @@
+      const response = await api.getQuizAPI();
+
+      // 파싱한 퀴즈 데이터 state에 저장
+      const parsedData = await quizParser(response.data);
```
여기 `await` 을 사용하신 이유는 무엇인가요??

jinseoplee replied:
API Call 은 promise 리턴해서 사용하였는데 그 외에는 필요가 없겠네요. 계속 수정하다가 소스를 정리하지 않았던것 같습니다. 불필요한 await 제거하였습니다.
fc4cd5f6e93096f748e9acb8a5baa2982bafe330

songbly commented on quiz/src/components/Training.vue:
```diff
@@ -0,0 +1,255 @@
+
+<style lang="scss" scoped>
+.training {
+  height: 100vh;
```
이것은 왜 100vh인가염???

jinseoplee replied:
grid 가용 면적을 뷰포트 전체로 잡고싶어서 100vh를 사용하였습니다

joointhezoo commented on quiz/src/components/Training.vue:
```diff
@@ -0,0 +1,255 @@
+    review() {
+      this.showReview = true;
+    },
+    restart() {
```
다시 풀기 되었을 때 default로 정답이 선택되어 있어, 이미 선택된 번호를 정답으로 선택하면 넘어가지지 않습니다. 💦

joointhezoo replied:
정답 확인 후, 풀었던 문제에 대한 선택된 정답이 바뀌네요. 사용자는 어떤 정답을 선택해서 틀렸는지 알고 싶을거 같아요. 🙋🏻‍♀️🔍

![May-07-2019 11-38-40](https://user-images.githubusercontent.com/12146287/57268181-d66b5900-70bd-11e9-8a78-244887b8b094.gif)

jinseoplee replied:
오류 수정하였습니다!
9863ae48030bae52d4c8daba9b8b0fde3e564746

shKwonRiiid commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+const extract = {
+  chunkMap: {},
+  blank() {
+  },
```
이 함수는 쓰이는 곳이 없는거 같네요 ✋

jinseoplee replied:
앗.. name이 blank인 데이터를 사용할지말지 고민하다가 그대로 둔 것 같습니다. 사용하지 않는 함수 제거하였습니다!
be27a04a0d264391dff908f6dfa246757f864c75

juyeong1260 commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+    }
+    return paragraphArray;
+  },
+  choice_area(ca) {
```
`question_area`, `choice_area` 가 같은 역할을 하는 함수로 보이는데 분리하신 이유가 있으신가요??

jinseoplee replied:
처음에 각 데이터의 name에 따라서 함수를 모두 생성하다 보니까 같은기능이 반복되는 경우가 많이 생겼습니다.😂반복되는 부분을 분리하여 수정하였습니다.

2d7cdc96799ede448ab8ad4663fc57ca54355187

shKwonRiiid commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+        choiceArray.push(this[child.name](child));
+      }
+    }
+    return choiceArray;
```
전체적으로 
`map`, `filter` 보단
새로운 배열에 push 하는 형태를 선호하시는거 같은데
이유가 있으신가요?

jinseoplee replied:
음 push를 선호한다기 보다는 그냥 익숙한 방법으로 사용했던거 같습니다.
map, filter 를 이용해 리펙토링 하면 가독성이 더 좋아질것 같아서 수정하였습니다.
2d7cdc96799ede448ab8ad4663fc57ca54355187

juyeong1260 commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+            result.passageTranslations.push(passageTranslationsParagraph);
+          }
+        }
+      }
```
요 위에 루프부분 굉장히 많이 중복되는 부분이 보이네요! 혹시 별도의 함수로 빼지 않고 위와같이 작성하신 이유가 있으신가요??

jinseoplee replied:
그게 더 가독성이 좋은듯 했으나.. 함수로 분리하니 훨씬 깔끔하네요. 위 항목과 같이 반복되는 부분 분리하여 적용하였습니다!
2d7cdc96799ede448ab8ad4663fc57ca54355187

beautifulife commented on quiz/src/components/Training.vue:
```diff
@@ -0,0 +1,255 @@
+        <p class="training__review" @click="review" v-else-if="isCompleted">리뷰</p>
+      </div>
+      <div class="training__quiz" v-if="!showReview">
+        <quiz-navigation class="training__quizNavigation" :count="questions.length" v-model="currentQuizIndex" :callback="quizNavigationClickCallback"/>
```
New Line(줄바꿈) 관련하여 어떠한 기준을 가지고 계신가요?

jinseoplee replied:
음.. block, inline 단위로 모두 줄바꿈을 해주고 있습니다! (이 부분에 대해서 여쭤보신게 맞는지..)

zittto committed:
eslint 설정 변경

beautifulife commented on quiz/config/dev.env.js:
```diff
@@ -0,0 +1,7 @@
+
+module.exports = merge(prodEnv, {
+  NODE_ENV: '"development"'
+})
```
lint가 적용되지 않도록 설정되어서인지,  `quiz/config/index.js` 와 다르게 `quiz/config/dev.env.js`, `quiz/config/prod.env.js`는 코드 스타일이 다르게 되어있습니다.

beautifulife replied:
유사한 상황이 `quiz/build` 에도 많이 있는데, 자동화 세팅으로 인한 이슈일까요?

jinseoplee replied:
아 그게 처음에 vue-cli로 프로젝트 세팅시 lint 설정을 하면 /build/, /config/ 경로를 아예 ignore 시킨 상태로 설정이 되더라구요.. 저는 그대로 사용했었는데 아무래도 유지보수를 위해서 컨벤션을 맞추는게 좋겠네요!!

jinseoplee replied:
/build/, /config/ 모두 린트 적용하였습니다!
e3e9725b085c08b0afaf0629107895b545172d75

zittto committed:
함수명 재정의, blank 함수 제거, destructuring 적용

1. QuizParser 함수 명 camel case 적용
  - QuizParser -> quizParser
2. quizParser 매개변수 destructuring 적용
  - const QuizParser(quiz) -> const quizParser({ name, type, preview, questions, chunk_map, passage_box })

3. extract.blank 함수 제거

beautifulife commented on quiz/src/utils/quiz-parser.js:
```diff
@@ -0,0 +1,156 @@
+  };
+};
+
+export default QuizParser;
```
`quiz/src/utils/quiz-parser.js` 작성하시면서 리팩토링 하고 싶으신 부분이 있으신지, 혹시 지금 변경하신다면 어떠한 방법으로 하실 생각이신지 궁금합니다.  ❓

jinseoplee replied:
데이터를 패턴화 해서 반복되는 부분을 좀 더 세밀하게 추상화 하고 싶습니다.

zittto committed:
리팩토링

1. 반복되는 부분 함수로 처리
2. for, push -> map, reduce, filter로 변경

zittto committed:
오류 수정

1. 다시 풀기 되었을 때 default로 정답이 선택되어 있는 문제 해결
2. 정답 확인 후 답 변경 못하도록 수정

zittto committed:
/build/, /config/ lint 적용

zittto committed:
action의 불필요한 await 제거
